cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -684,6 +685,85 @@ Result<S3Model::GetObjectResult> GetObjectRange(Aws::S3::S3Client* client,
   return OutcomeToResult(client->GetObject(req));
 }
 
+template <typename ObjectResult>
+std::shared_ptr<const KeyValueMetadata> GetObjectMetadata(const ObjectResult& result) {
+  std::vector<std::string> keys, values;
+
+  auto push = [&](std::string k, const Aws::String& v) {
+    if (!v.empty()) {
+      keys.push_back(std::move(k));
+      values.push_back(FromAwsString(v).to_string());
+    }
+  };
+  auto push_datetime = [&](std::string k, const Aws::Utils::DateTime& v) {
+    if (v != Aws::Utils::DateTime(0.0)) {
```

Comment:
I suppose this is unlikely to be a valid "Last-Modified" or "Expires" but it still looks odd

Reply:
Well, this is the value returned by the AWS SDK when the information is unavailable, AFAIK.